src/Rings/mpoly-nested.jl

```diff
@@ -0,0 +1,299 @@
+##############################################################################
+#
+# Various special cases for nested rings. Currently:
+#  1. frac(R[t1, t2, ...])[x1, x2, ...]
+#  2. R[z1, ...]...[y1, ...][x1, ...] and univariate possibilities
+#
+##############################################################################
+
+export denest, renest
+
+##############################################################################
+#
+# 1. implementations for Q(t)[x]
+#
+##############################################################################
+
+# from Q(t1, t2)[x1, x2] to S = Q[x1, x2, t1, t2]
+# the product of the two outputs should be equal to the input
+function _remove_denominators(S::MPolyRing, f::MPolyElem)
+  R = parent(f)                   # Q(t1, t2)[x1, x2]
+  R2 = base_ring(base_ring(R))    # Q[t1, t2]
+  den = one(R2)
+  cont = zero(R2)
+  for (c1, e1) in zip(coefficients(f), exponent_vectors(f))
+    cont = gcd(cont, numerator(c1))
+    den = lcm(den, denominator(c1))
+  end
+  num = MPolyBuildCtx(S)
+  for (c1, e1) in zip(coefficients(f), exponent_vectors(f))
+    cf = divexact(den, denominator(c1))
+    if !iszero(cont)
+      cf *= divexact(numerator(c1), cont)
+    end
+    for (c2, e2) in zip(coefficients(cf), exponent_vectors(cf))
+      push_term!(num, c2, vcat(e1, e2))
+    end
+  end
+  return (finish(num), cont//den)
+end
+
+# from Q[x1, x2, t1, t2] to R = Q(t1, t2)[x1, x2]
+# the output is equal to the input
+function _restore_numerators(R::MPolyRing, g::MPolyElem)
+  S = parent(g)                   # Q[x1, x2, t1, t2]
+  R2 = base_ring(base_ring(R))    # Q[t1, t2]
+  n = nvars(R)
+  d = Dict{Vector{Int}, typeof(MPolyBuildCtx(R2))}()
+  for (c, e) in zip(coefficients(g), exponent_vectors(g))
+    e1 = e[1:n]
+    e2 = e[n+1:end]
+    if !haskey(d, e1)
+      d[e1] = MPolyBuildCtx(R2)
+    end
+    push_term!(d[e1], c, e2)
+  end
+  z = MPolyBuildCtx(R)
+  for (e, c) in d
+    push_term!(z, base_ring(R)(finish(c)), e)
+  end
+  return finish(z)
+end
+
+# if R2 is a singular ring, keep it that way since limited types can be factored
+function _add_variables(R2::Singular.PolyRing, n::Int)
+  n += nvars(R2)
+  x = map(i->"x"*string(i), 1:n)
+  return Singular.PolynomialRing(base_ring(R2), x, cached=false)[1]
+end
```

Comment:
It seems this function is not tested? At least according to codecov.

Reply:
Or maybe codecov did not properly?

Reply:
This will be tested when singular.jl is updated.

https://github.com/oscar-system/Oscar.jl/pull/421/files#diff-abcf02b5ef901a4fb66e89559dc89c77234bc2051c462af74c9fa93fb7afc635R52